servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -0,0 +1,432 @@
+      } finally {
+        // Downgrade to read lock again after eviction
+        mEvictionLock.readLock().lock();
+        mEvictionLock.writeLock().unlock();
+      }
+      tempBlock = mAllocator.allocateBlock(userId, blockId, initialBlockSize, location);
+      Preconditions.checkNotNull(tempBlock, "Cannot allocate block %s:", blockId);
+    }
+    // Add allocated temp block to metadata manager
+    mMetaManager.addTempBlockMeta(tempBlock);
+    return tempBlock;
+  }
+
+  private void commitBlockNoLock(long userId, long blockId, TempBlockMeta tempBlockMeta)
+      throws IOException {
+    if (mMetaManager.hasBlockMeta(blockId)) {
+      throw new IOException("Failed to commit block " + blockId + ": block is committed");
+    }
+    // Check the userId is the owner of this temp block
+    long ownerUserId = tempBlockMeta.getUserId();
+    if (ownerUserId != userId) {
+      throw new IOException("Failed to commit temp block " + blockId + ": ownerUserId "
+          + ownerUserId + " but userId " + userId);
+    }
+    String sourcePath = tempBlockMeta.getPath();
+    String destPath = tempBlockMeta.getCommitPath();
+    boolean renamed = new File(sourcePath).renameTo(new File(destPath));
+    if (!renamed) {
+      throw new IOException("Failed to commit temp block " + blockId + ": cannot rename from "
+          + sourcePath + " to " + destPath);
+    }
+    mMetaManager.commitTempBlockMeta(tempBlockMeta);
+  }
+
+  private void abortBlockNoLock(long userId, long blockId) throws IOException {
+    if (mMetaManager.hasBlockMeta(blockId)) {
+      throw new IOException("Failed to abort block " + blockId + ": block is committed");
+    }
+
+    TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
+    // Check the userId is the owner of this temp block
+    long ownerUserId = tempBlockMeta.getUserId();
+    if (ownerUserId != userId) {
+      throw new IOException("Failed to abort temp block " + blockId + ": ownerUserId "
+          + ownerUserId + " but userId " + userId);
+    }
+    String path = tempBlockMeta.getPath();
+    boolean deleted = new File(path).delete();
+    if (!deleted) {
+      throw new IOException("Failed to abort temp block " + blockId + ": cannot delete " + path);
+    }
+    mMetaManager.abortTempBlockMeta(tempBlockMeta);
+  }
+
+  private void moveBlockNoLock(long blockId, BlockStoreLocation newLocation) throws IOException {
+    if (mMetaManager.hasTempBlockMeta(blockId)) {
+      throw new IOException("Failed to move block " + blockId + ": block is uncommited");
+    }
+    BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
+    String srcPath = blockMeta.getPath();
+    blockMeta = mMetaManager.moveBlockMeta(blockMeta, newLocation);
+    String destPath = blockMeta.getPath();
+
+    if (!new File(srcPath).renameTo(new File(destPath))) {
+      throw new IOException("Failed to move block " + blockId + ": cannot rename from " + srcPath
+          + " to " + destPath);
+    }
```

Comment:
It seems like in the other methods, we do the data operation before the metadata operation, is there any reason we do it differently here? And if an error occurs, should the metadata operation be reverted?